src/bun.js/api/bun/subprocess.zig

```diff
@@ -2483,6 +2500,23 @@ pub const Subprocess = struct {
             _ = maybe catch |err| return globalThis.handleError(err, "in configuring child stderr");
         }
 
+        for (stdio_inherits.items) |fileno| {
+            const maybe = blk: {
+                if (comptime Environment.isMac) {
+                    actions.inherit(bun.toFD(fileno)) catch |err| break :blk err;
+                } else {
+                    actions.dup2(bun.toFD(fileno), bun.toFD(fileno)) catch |err| break :blk err;
+                }
+            };
+            _ = maybe catch |err| return globalThis.handleError(err, "in configuring child stdio");
+        }
+        stdio_inherits.deinit(bun.default_allocator);
```

Comment:
I'd prefer to move the deinit to a `defer` block (like the one around line 1984). That way, an error exit (in, say, line 2511) won't leak memory.

Reply:
> Thanks for that. Pushed a commit with the defers.

Note that we might want to use `allocator` and skip the deinit for those variables. I'm not really sure about the tradeoffs, to be honest.

> This raises another question: doesn't stdio_pipes have the same problem? (Fixing it for stdio_pipes is probably out of scope for this PR, but good to have eyes looking at this.)

I believe it does, but we should have a separate PR (or separate PRs) to fix the remaining problems with spawnMaybeSync:
1. deinit stdio_pipes on error (but not, of course, on success).
2. remove the "leak?" comment, after ensuring there's no leak left
3. allow empty arguments (rare, but useful, and potentially security-relevant as a maliciously empty string can shift option arguments into non-option slots)
4. set PATH before using it rather than afterwards
5. handle keys containing NUL characters in env and argv, which may otherwise result in very strange behavior
6. test all this

Reply:
If `allocator` is stack-scoped (w/ an escape hatch to use heap or some other fallback, if a limit is reached; like some of the polymorphic allocators for >=C++17), then I'm for it. I don't know the ( bun/zig ) code well enough to be aware, but it seems like a reasonable expectation.

Regarding the leak, is there a leak risk? I thought the defer should pretty much solve that.

To make the history clean, I propose to squash the commits with you mentioned as a "Co-authored-by:" ... sound okay?

Reply:
> Regarding the leak, is there a leak risk? I thought the defer should pretty much solve that.

I think the code is fine, but there's a comment there that says "leak?", which is confusing.

> To make the history clean, I propose to squash the commits with you mentioned as a "Co-authored-by:" ... sound okay?

Absolutely, feel free to do whatever makes the most sense to you.

Reply:
Oh right, I see what you are saying about the "leak?" comment now. I pushed a squashed version of the commits. So I guess we're set now?

Reply:
One moment .... local allocator commit coming soon.

Reply:
More questions :) What order do the defers execute in? Does the defer deinit of the local allocator (arena) happen before or after the things that use it? Also is there any point to doing a deinit of Arrays that allocate from it since they will (I assume) implicitly go away when the area deinits?

Reply:
> More questions :) What order do the defers execute in? Does the defer deinit of the local allocator (arena) happen before or after the things that use it?

defers execute in reverse/LIFO order (within a scope), so the arena is deinited after all the things that use it are.

> Also is there any point to doing a deinit of Arrays that allocate from it since they will (I assume) implicitly go away when the area deinits?

Certainly no point if the deinit is the last thing that happens before the arena deinit. The only reason I can see is if we allocated so much memory we might want to free it before allocating more, which isn't the case here.